Proposal: Swift Distributed Tracing (#84)
swift-server/sswg: proposals/NNNN-swift-distributed-tracing.md

internals. They implement generic tracing support using the Swift Distributed Tracing API without a specific tracing
backend in mind.

#### Examples

- [Hummingbird](https://github.com/hummingbird-project/hummingbird/releases/tag/1.6.0)
- [Soto](https://github.com/soto-project/soto-core/pull/575)

### Tracer backend implementations

The final piece of the puzzle are tracer backend implementations. They provide vendor-specific support for exporting
tracing spans.

#### Examples

[Swift OTel](https://github.com/slashmo/swift-otel) exposes a tracer that exports to
an [OpenTelemetry Collector](https://opentelemetry.io/docs/collector). This already allows adopters of this tracing
library to export to popular backends compatible with OpenTelemetry, such as Zipkin, Jaeger, Honeycomb, and more.

## Detailed design

Both library/framework authors and end-users use the same API to create tracing spans. Adding a span to functions can
be achieved by using the `withSpan(operationName:context:body:)` API:

```swift
func performOperation() -> String {
    withSpan("operation") {
        return "handled"
    }
}
```

### Customizing the span

> A span can be further customized using `withSpan(operationName:context:body:)` as well:

Reviewer: How is the span customised using `withSpan(operationName:context:body:)` in the example below. The text implies the `withSpan` function is doing the customisation. Also the function signature does not include `ofKind:`which you use in the example

Reviewer 2: Sorry I don't get this question:
> How is the span customised using withSpan(operationName:context:body:) in the example below.

> ```
>     withSpan(request.endpointPath, ofKind: .server) { span in
>         span.attributes["http.method"] = request.method.rawValue
> ```

It does that ^ but the primary purpose is to set the task-local. Should we explain this more?

Reviewer 2: I added: [`da9f679` (#84)](https://github.com/swift-server/sswg/pull/84/commits/da9f6790f53ed9f4162f173f04cd11b2bb9d04ab)

Reviewer: It is just the text doesn't align with the example. You imply it is the function `withSpan` that is doing the customisation, not the editing of the attributes `span.attributes["http.method] = ...`. Its not a huge issue

> A span can be further customized using `withSpan(operationName:context:body:)` as well:

Reviewer 2: This signature is a bit of a lie, we could use the full signature: `withSpan(_:context:ofKind:function:file:line:operation:)`

```suggestion
A span can be further customized using `withSpan(_:context:ofKind:function:file:line:operation:)` as well. In the example below we're adding a custom attribute (`"http.method"`) to the span, so when viewing it in a trace visualization backend we can have more detailed information about what kind of request this was:
```

    }
}
```

### Child Spans

A distributed tracing span may have one or more child spans. While Swift Distributed Tracing also allows for manually
linking spans, creating a child span is easiest when using the
task-local [`ServiceContext`](https://github.com/apple/swift-service-context) behind the scenes:

```swift
func handleRequest(_ request: Request) async throws -> Response {
    try await withSpan(request.endpointPath, ofKind: .server) {
        let user = try await fetchUser(byID: request.userID)
        return try await respond(to: request)
    }
}

func fetchUser(byID id: String) async throws -> User {
    try await withSpan("fetchUser") { span in
        span.attributes["user.id"] = id
    }
}
```

In this example, the `withSpan(operationName:context:body:)` call within `handleRequest(_:)` creates a span.
When creating the span, the tracer implementation sets trace and span identifiers on the task-local
[ServiceContext](https://github.com/apple/swift-service-context). `fetchUser(id:)` will then automatically "inherit"
these identifiers, indicating to the tracer implementation that the "fetchUser" span should be a child of the request
span.

### Low-level API

> When using Swift NIO `EventLoopFuture`s or similar patterns, we need to control the spans lifetime manually. This can be

Reviewer: Maybe mention the service context will need to be propagated manually if using EventLoopFutures.

Reviewer 2: ```suggestion
When using Swift NIO `EventLoopFuture`s or similar patterns, we need to control the spans lifetime manually. 
The service context will also need to be propagated manually by passing it to the context parameter of startSpan, 
because event loop futures cannot automatically propagate task local values.
This can be
```

```swift
func handleRequest(_ request: Request) async throws -> Response {
    try await withSpan(request.endpointPath, ofKind: .server) {
        let user = try await fetchUser(byID: request.userID)
        return try await respond(to: request)
    }
}

func fetchUser(byID id: String) async throws -> User {
    try await withSpan("fetchUser") { span in
        span.attributes["user.id"] = id
    }
}
```

In this example, the `withSpan(operationName:context:body:)` call within `handleRequest(_:)` creates a span.
When creating the span, the tracer implementation sets trace and span identifiers on the task-local
[ServiceContext](https://github.com/apple/swift-service-context). `fetchUser(id:)` will then automatically "inherit"
these identifiers, indicating to the tracer implementation that the "fetchUser" span should be a child of the request
span.

### Low-level API

When using Swift NIO `EventLoopFuture`s or similar patterns, we need to control the spans lifetime manually. 
The service context will also need to be propagated manually by passing it to the context parameter of startSpan, 

> because event loop futures cannot automatically propagate task local values.

Reviewer: Maybe the example needs to show this
```swift
func handleRequest(_ request: Request, context: ServiceContext)
...
```

Reviewer 2: done! good catch that really helps the example, I added some comments too

achieved using the lower-level `startSpan(operationName:context:)` API:

```swift
func handleRequest(_ request: Request, context: ServiceContext) -> EventLoopFuture<Response> {
    // Start the span manually with startSpan passing in the `context` explicitly 
    let span = startSpan(request.endpointPath, context: context, ofKind: .server)

    // Chain an "always" onto the response future since we must ALWAYS end() a span
    return response(to: request).always { result in
        switch result {
        case .success(let response):
            span.attributes["http.status_code"] = response.status.code
        case .failure(let error):

>             // If the response has failed, mark this span as failed by recording an error

Author: ```suggestion
            // If the response has failed, add the error to the span by recording it.
```

Small nitpick: I wouldn't call it "mark the span as failed" since that's what the span status is used for. Recording an error doesn't necessarily mean that the span failed, e.g., if you're expecting an unsuccessful status code.

Reviewer 2: Thanks, true!
